pyfr/nputil.py

```diff
@@ -126,3 +127,187 @@ def npdtype_to_ctypestype(dtype):
         return None
 
     return _ctypestype_map[np.dtype(dtype).type]
+
+
+#Get indices for interval segments of interpolation
+def get_interval_idxs(t, xarr):
+    if np.asarray(xarr).size > 1:
+        idxs = np.array([np.max(np.nonzero(xx >= t[:-1] - 1e-12)[0]) 
+                                                        for xx in xarr])
+    else:
+        idxs = np.max(np.nonzero(xarr >= t[:-1] - 1e-12)[0])
+    return idxs
+
+#Modified Gauss Elimination/Thompson algorithm for tridiag systems
+#All changes are done in place and the algorithm 
+#is applied to the lhs and rhs together
+def trislv(a, b, c, d):
+    #forward elimination
+    d[0] /= b[0]
+    for i, (ai, bi, bj, ci, dj) in enumerate(
+                zip(a, b[1: ], b[: -1], c, d[: -1]), start=1):
+        #lhs coeffs
+        c[i-1] = cdivb = ci/bj
+        b[i] = bi1 = bi - ai*cdivb
+        #rhs     
+        d[i] -= ai*dj
+        d[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    for i, (ci, dj) in enumerate(zip(c[::-1], d[:0:-1]), start=1):
+        dd[i] -= ci*dj
+
+
+#tridiagonal solver for cyclic/periodic systems 
+#Using Sherman-Morrison formula and Thompson algorithm
+# A = A + u ⊗ v
+# u, v are vectors
+def trislvcyc(a, b, c, d, alpha, beta):
+    u = np.zeros(d.shape[0], dtype=d.dtype)
+    #Solve the two systems, Ax = d, Bz = u in place, i.e., d->x, u->z
+    gam = -b[0] 
+    b[0] *= 2
+    b[-1] -= alpha*beta/gam
+    u[0] = gam
+    u[-1] = alpha
+    #solve the two systems in one loop
+    #forward elimination
+    d[0] /= b[0]
+    u[0] /= b[0]
+    for i, (ai, bi, bj, ci, dj, uj) in enumerate(
+                zip(a, b[1: ], b[: -1], c, d[: -1], u[: -1]), start=1):
+        #lhs coeffs
+        c[i-1] = cdivb = ci/bj
+        b[i]  = bi1 = bi - ai*cdivb  
+        # first rhs
+        d[i] -= ai*dj
+        d[i] /= bi1
+        # second rhs
+        u[i] -= ai*uj
+        u[i] /= bi1
+
+    #backward substitution
+    dd = d[::-1]
+    uu = u[::-1]
+    for i, (ci, dj, uj) in enumerate(
+            zip(c[::-1], d[:0:-1], u[:0:-1]), start=1):
+        dd[i] -= ci*dj
+        uu[i] -= ci*uj
+
+    #compute the final solution
+    fact = (d[0] + alpha*d[-1]/gam)/(1.0 + u[0] + alpha*u[-1]/gam)
+    d -= fact*u[:, None]
+
+    
+class CurveFit:
```

Comment:
This class is pointless.

Reply:
The point is to encapsulate different types of interpolation, linear and spline, etc.

Reply:
But it encapsulates nothing!  There is no common code factored out.

Reply:
It makes it easy to always just call Class.interp without worrying about which interp type is used in the fwh solver. It is just decided at initialization if linear or others. The CurveFit is also made available for future development if needed to add more and have common parts. Since I am just offering linear and cubic spline for now there is no common code in base class.

Reply:
Sure, but that does not require a class hierarchy.  Python is duck typed.  You are free to call any method on any object you like and so long as a particular object implements the method in question everything will work.

Developing something now with the view that it will be of utility in the future is seldom a sensible proposition.  You're imposing additional complexity now (for everyone reading the code) in the hope that it will be somehow beneficial in the future.  It is much better to keep things lightweight and only later, when more specifics are known, try and factor out common code.

Reply:
I started with it as a bunch of functions but then thought this can be helpful. Anyway, I will convert all to just functions unless you have another suggestion.

Reply:
If you need state it is best to have an object which overloads `__call__`.  Then if no state is needed a bare function can work, and if state is needed then you can create an instance of an object which can be used as if it were a function.

Reply:
Done, let me know if you have other ideas.